controllers/reconcilers/kserve_isvc_gateway_reconciler.go

```diff
@@ -0,0 +1,263 @@
+	istioclientv1beta1 "istio.io/client-go/pkg/apis/networking/v1beta1"
+)
+
+const (
+	internalSerivceHostnameDomain = "svc.cluster.local"
+	opendatahubManagedLabelName   = "opendatahub.io/managed='true'"
+	secretCheckInterval           = 1 * time.Second
+	maxSecretCheckAttempts        = 5
+)
+
+var _ SubResourceReconciler = (*KserveGatewayReconciler)(nil)
+
+type KserveGatewayReconciler struct {
+	client         client.Client
+	secretHandler  resources.SecretHandler
+	gatewayHandler resources.GatewayHandler
+	deltaProcessor processors.DeltaProcessor
+}
+
+func NewKserveGatewayReconciler(client client.Client) *KserveGatewayReconciler {
+	return &KserveGatewayReconciler{
+		client:         client,
+		secretHandler:  resources.NewSecretHandler(client),
+		gatewayHandler: resources.NewGatewayHandler(client),
+		deltaProcessor: processors.NewDeltaProcessor(),
+	}
+}
+
+func (r *KserveGatewayReconciler) Reconcile(ctx context.Context, log logr.Logger, isvc *kservev1beta1.InferenceService) error {
+	log.V(1).Info("Reconciling KServe local gateway for Kserve InferenceService")
+
+	// return if URL is not set
+	if isvc.Status.URL == nil {
+		log.V(1).Info("Waiting for the URL as the Inference Service is not ready yet")
+		return nil
+	}
+
+	destinationSecretExist := true
+	if _, err := r.secretHandler.Get(ctx, types.NamespacedName{Name: isvc.Name, Namespace: constants.IstioNamespace}); err != nil {
+		if errors.IsNotFound(err) {
+			destinationSecretExist = false
+		} else {
+			return err
+		}
+	}
+
+	if !destinationSecretExist {
+		secret := &corev1.Secret{}
+		for attempt := 1; attempt <= maxSecretCheckAttempts; attempt++ {
+			getSecretErr := r.client.Get(ctx, types.NamespacedName{Name: isvc.Name, Namespace: isvc.Namespace}, secret)
+			if getSecretErr == nil {
+				break
+			}
+			if errors.IsNotFound(getSecretErr) {
+				log.V(2).Info("The certificate secret is not created yet. Retrying...", "attempt_number", attempt)
+
+				time.Sleep(secretCheckInterval)
+			} else {
+				log.Error(getSecretErr, "Failed to retrieve the certificate secret for the InferenceService (ISVC)")
+				return getSecretErr
+			}
+		}
+		if err := r.copyServingCertSecretFromIsvcNamespace(ctx, secret); err != nil {
+			log.V(1).Error(err, "Failed to update KServe local gateway in the istio-system namespace")
+
+			return nil
+		}
+	}
+
+	// Create Desired resource
+	desiredResource, err := r.createDesiredResource(isvc)
+	if err != nil {
+		return err
+	}
+
+	// Get Existing resource
+	existingResource, err := r.getExistingResource(ctx)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			log.Error(err, "Failed to find KServe local gateway in istio-system namespace")
+		}
+		return err
+	}
+
+	// Process Delta
+	if err = r.processDelta(ctx, log, desiredResource, existingResource); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *KserveGatewayReconciler) createDesiredResource(isvc *kservev1beta1.InferenceService) (*istioclientv1beta1.Gateway, error) {
```

Comment:
the verb create led us to think that is it being created/deployed in the cluster, maybe `getDesiredResource`?

Reply:
It is kind of standard format that @vaibhavjainwiz defined for odh-model-controller. So I followed the way and I think the name would be ok.

Reply:
just my 2 cents, we could change that :D
it is actually not creating anything :)

Reply:
yes. I agreed.
updated.